velox/docs/develop/spilling.rst

```diff
@@ -0,0 +1,480 @@
+spill on all the operators (also called as group spill in discussion below). It
+will first collect spillable stats from all the operators and then select a
+number of partitions to spill.
+
+.. code-block:: c++
+
+    void Spiller::spill(const SpillPartitionNumSet& partitions);
+
+**spill vector**: the operator spills a row vector to a specified partition. The
+Spiller directly appends the row vector to the currently open spill file from
+that partition. The spilling process is also controlled by the operator in this
+case. It is used for spilling by the hash join. Both hash build and hash probe
+operators will spill input rows to disk if the corresponding partition has been
+spilled. For the hash build operator, if a partition has been spilled, then all
+the input rows from that partition have to spill as we can’t build a hash table
+with a subset of rows from that partition to join. For the hash probe operator,
+it itself is not spillable but it needs to spill the input rows if the
+associated partition has been spilled by the hash build. We will discuss this
+further in the hash join spilling section.
+
+.. code-block:: c++
+
+    void Spiller::spill(uint32_t partition, const RowVectorPtr& spillVector);
+
+`SpillFileList <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L155>`_ and `SpillFile <https://github.com/facebookincubator/velox/blob/main/velox/exec/Spill.h#L58>`_
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+There is one SpillFileList object for each spill partition to manage its spill
+files. Each spill file is managed by one SpillFile object which provides the
+low level io operations with the storage system through Velox file system
+interface. On the spill path, the SpillFileList object takes a row vector as
+input, and internally it creates a VectorStreamGroup to serialize the row
+vector and write out the serialized byte stream into the currently opened spill
+file through the corresponding SpillFile object. The SpillFileList object will
+also switch to the next spill file if the current file grows beyond the target
+file size. On the restore path, the SpillFile object reads the serialized byte
+stream from the underlying storage system, and then uses VectorStreamGroup to
+deserialize the byte stream into row vectors.
+
+Spill Triggers
+--------------
+
+Spilling will be integrated with the Velox memory management system as a way to
+reclaim memory when the system lacks memory. Whenever an operator can’t
+allocate or reserve new memory, the memory arbitrator will select a number of
+Velox tasks to shrink their memory usage to satisfy the new memory allocation
+or reservation request. Each selected task will try to reclaim memory from its
+spillable operators. The latter in turn frees up memory by spilling out (part)
+of its memory state to disk. The integration of spilling with the memory
+management system is being under development.
+
+We also support to trigger spilling if the spillable operator's memory usage
```

Comment:
> We also support to trigger spilling

Velox can be configured to trigger spilling ...

It would be nice to describe the configuration properties that can be used to achieve that.